### PR TITLE
Add gitlab support

### LINKED ISSUE #191: How can I support Gitlab CI?
We need to support gitlab ci instead of circle ci.

### PR DESCRIPTION
**Pull request by agamm**
Fixes #191

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -303,6 +303,20 @@ default value for `-repotoken`.
 
 You can use the `-v` flag to see verbose output.
 
+
+## Gitlab CI
+
+Store your Coveralls API token as an [Environment Variable](https://docs.gitlab.com/ee/ci/variables/#create-a-custom-variable-in-the-ui).
+
+```
+COVERALLS_TOKEN=your_token_goes_here
```
Comment (gdm85) [inline README.md:312]: Can you remove this? The variable is defined through the GitLab UI, not declared like this.

File: README.md [hunk 2]
```diff
@@ -303,6 +303,20 @@ default value for `-repotoken`.
...
+```
+
+```
+$ go get github.com/mattn/goveralls
+$ goveralls -service=gitlab -repotoken=$COVERALLS_TOKEN
```
Comment (gdm85) [inline README.md:317]: If the environment variable is `COVERALLS_TOKEN` then it is not needed to pass `-repotoken`, which is in fact insecure.

Also, wouldn't it be better to provide here a YAML snippet instead?

File: goveralls.go [hunk 3]
```diff
@@ -407,6 +409,8 @@ func process() error {
 		ghPR := githubEvent["pull_request"].(map[string]interface{})
 		ghHead := ghPR["head"].(map[string]interface{})
 		head = ghHead["sha"].(string)
+	} else if prNumber := os.Getenv("CI_EXTERNAL_PULL_REQUEST_IID"); prNumber != "" {
+		pullRequest = prNumber
```
Comment (gdm85) [inline goveralls.go:413]:
```suggestion
		// pull request id from GitHub when building on GitLab
		pullRequest = prNumber
```

**Review by gdm85 — CHANGES_REQUESTED** [review]
Thanks for submitting this PR!

I am requesting some changes here and also I think the `CI_MERGE_REQUEST_ID` or `CI_MERGE_REQUEST_IID` should be covered to get the Merge Request ID (corresponding to PR number)

File: README.md (hunk 2 above)
Comment (agamm) [reply README.md]: Better?

**Comment by agamm** [comment]
> Thanks for submitting this PR!
> 
> I am requesting some changes here and also I think the `CI_MERGE_REQUEST_ID` or `CI_MERGE_REQUEST_IID` should be covered to get the Merge Request ID (corresponding to PR number)

I didn't really understand what you meant by: `CI_MERGE_REQUEST_IID`, did my changes relate? Also I didn't find where to test this part of the code.

File: README.md [hunk 4]
```diff
@@ -303,6 +303,28 @@ default value for `-repotoken`.
...
+Store your Coveralls API token as an [Environment Variable](https://docs.gitlab.com/ee/ci/variables/#create-a-custom-variable-in-the-ui).
+
+
+```yml
+publish:coverage:
+  timeout: 30m
+  stage: publish
+  artifacts:
+    paths:
+      - covprofile
```
Comment (gdm85) [inline README.md:318]: Let's stick to the same filename used in this README
```suggestion
      - coverage.txt
```

File: README.md [hunk 5]
```diff
@@ -303,6 +303,28 @@ default value for `-repotoken`.
...
+  dependencies:
+    - build:env
+  when: always
+  script:
+    - go test -covermode atomic -coverprofile=covprofile ./...
```
Comment (gdm85) [inline README.md:323]:
```suggestion
    - go test -covermode atomic -coverprofile=coverage.txt ./...
```

File: README.md [hunk 6]
```diff
@@ -303,6 +303,28 @@ default value for `-repotoken`.
...
+    - go get github.com/mattn/goveralls
+    - goveralls -service=gitlab -coverprofile=covprofile
```
Comment (gdm85) [inline README.md:325]:
```suggestion
    - goveralls -service=gitlab -coverprofile=coverage.txt
```

File: README.md [hunk 7]
```diff
@@ -303,6 +303,28 @@ default value for `-repotoken`.
 
 You can use the `-v` flag to see verbose output.
 
+
+## Gitlab CI
+
+Store your Coveralls API token as an [Environment Variable](https://docs.gitlab.com/ee/ci/variables/#create-a-custom-variable-in-the-ui).
+
+
```
Comment (gdm85) [inline README.md:311]:
```suggestion
```

**Review by gdm85 — CHANGES_REQUESTED** [review]

**Comment by gdm85** [comment]
> I didn't really understand what you meant by: `CI_MERGE_REQUEST_IID`, did my changes relate? Also I didn't find where to test this part of the code.

Your PR introduces use of `CI_EXTERNAL_PULL_REQUEST_IID` for the PR number; that is only valid when building a GitHub project on GitLab. If you are building a MR on GitLab using a GitLab project, then you need either `CI_MERGE_REQUEST_ID` or `CI_MERGE_REQUEST_IID` (I suspect the latter).

Can you please verify?

**Comment by agamm** [comment]
> CI_MERGE_REQUEST_ID

I work by mirroring my github to gitlab so I can't really verify that super easily. What did you have in mind, I'll do my best?
PS it seems that `CI_MERGE_REQUEST_ID` is more relevant. ("unique ID across all projects on GitLab")

**Comment by agamm** [comment]
I created a temporary repo to test this fork, I always got an ambigious message of a `500 error`:
> Bad response status from coveralls: 500
> {"message":"Build processing error.","error":true,"url":""}

Pipeline example: https://gitlab.com/agmore1/testgoveralls/-/jobs/850406137
.gitlab-ci.yml: https://gitlab.com/agmore1/testgoveralls/-/blob/master/.gitlab-ci.yml

**Comment by gdm85** [comment]
> I created a temporary repo to test this fork, I always got an ambigious message of a `500 error`:
> 
> > Bad response status from coveralls: 500
> > {"message":"Build processing error.","error":true,"url":""}
> 
> Pipeline example: https://gitlab.com/agmore1/testgoveralls/-/jobs/850406137
> .gitlab-ci.yml: https://gitlab.com/agmore1/testgoveralls/-/blob/master/.gitlab-ci.yml

That sounds like a bug in coveralls.io itself; cc @nickmerwin

@agamm please update your project to use `go get github.com/mattn/goveralls@master` and then run it with with `-debug=true`, that might provide interesting clues for troubleshooting the server-side bug.

> I work by mirroring my github to gitlab so I can't really verify that super easily. What did you have in mind, I'll do my best?
> PS it seems that `CI_MERGE_REQUEST_ID` is more relevant. ("unique ID across all projects on GitLab")

I think what we need here is the MR ID, so that coveralls.io can link back to GitLab with the right reference. Similarly, for GitHub it's the PR number not the unique PR ID across all projects.

**Comment by agamm** [comment]
> -debug=true

I tried updating from master but there were no new changes, I did change the merge request id to `iid` like you said, although I still honestly don't understand that part.

Seems to still fail:
https://gitlab.com/agmore1/testgoveralls/-/jobs/860029893

**Comment by agamm** [comment]
@gdm85 any update?

**Review by gdm85 — APPROVED** [review]

**Comment by gdm85** [comment]
@agamm that `500` is a bug on coveralls.io side; you can try using `-shallow` but they need to fix it on their side.

**Comment by afinetooth** [comment]
FYI @agamm and @gdm85, on behalf of coveralls, we provided a workaround for the current issue that entails adding the `commit-sha` to the base of your posted json. However, a fix will go to production in 24-48 hrs so please try again and let us know if it still occurs. thanks.

**Pull request merged**